### PR TITLE
feat: Implement incremental update protocol with multi-version compatibility

### PR DESCRIPTION
**Pull request by mingri31164**
## **Problem**

1. **Protocol rigidity**: Field renaming causes continuous invalid refreshes due to MD5 mismatch
2. **Poor extensibility**: Adding extended parameters triggers unnecessary client refreshes

## **Solution**

1. Multi-version protocol support (v1=full MD5, v2+=core MD5)
2. Segment-level comparison algorithm: separate core and extended parameters
3. Field adapter pattern for backward compatibility (`coreSize`/`corePoolSize`)
4. Version-aware MD5 comparison on server side

## **Key Changes**

1. Add `IncrementalContentUtil` for core content extraction
2. Add `IncrementalMd5Util` for versioned MD5 generation
3. Add field adapters in `ThreadPoolParameterInfo` (`corePoolSizeAdapt()`, `maximumPoolSizeAdapt()`)
4. Update `ConfigCacheService` for version-aware comparison
5. Client sends protocol version via `X-Hippo4j-Protocol-Version` header

## **Test Coverage**

1. `ProtocolRigidityTest`: 5 scenarios for field renaming compatibility
2. `ExtensibilityTest`: 6 scenarios for extended parameter addition
3. `ContentUtilTest`: 4 scenarios for field adapter verification

---

### CODE REVIEW

File: dependencies/pom.xml [hunk 1]
```diff
@@ -112,6 +113,16 @@
                 <version>${h2.version}</version>
                 <scope>runtime</scope>
             </dependency>
+            <dependency>
```
Comment (magestacks) [inline dependencies/pom.xml]: Why introduce the ByteBuddy framework?

File: dependencies/pom.xml [hunk 2]
```diff
@@ -17,7 +17,8 @@
         <jjwt.version>0.9.0</jjwt.version>
         <jackson-bom.version>2.11.1</jackson-bom.version>
         <h2.version>2.1.214</h2.version>
-        <mockito.version>3.12.4</mockito.version>
+        <mockito.version>4.11.0</mockito.version>
```
Comment (magestacks) [inline dependencies/pom.xml]: Why was the version number changed? And why was the mockito-inline dependency removed below?

**Comment by magestacks** [comment]
@mingri31164 Please describe your testing process for the submitted code.

Also, does your submitted code cover all scenarios? For example, does it support smooth compatibility between different versions such as server 2.0.0 and client 1.5.0?

**Comment by mingri31164** [comment]
Hi @magestacks , thank you for the review! Here's a comprehensive overview of my testing approach:

### 1️⃣ Unit Testing

#### 1. **ProtocolRigidityTest** (5 test cases)

Addresses the field renaming issue:

| Scenario                                                     | Result                      |
| ------------------------------------------------------------ | --------------------------- |
| Old client (`coreSize/maxSize`) + New server (`corePoolSize/maximumPoolSize`) | ✅ MD5 identical, no refresh |
| Field adapter priority (new fields over old)                 | ✅ Correct priority          |
| Field adapter fallback (old fields when new missing)         | ✅ Correct fallback          |
| v1 vs v2 protocol difference                                 | ✅ Different strategies      |

#### 2. **ExtensibilityTest** (6 test cases)

Addresses the extended parameter issue:

| Scenario                                        | Result                     |
| ----------------------------------------------- | -------------------------- |
| Server adds `executeTimeOut` extended parameter | ✅ v2 client MD5 unchanged  |
| Server adds multiple extended parameters        | ✅ v2 client MD5 unchanged  |
| Only extended params changed                    | ✅ `hasCoreChanges=false`   |
| v1 vs v2 behavior on extended param changes     | ✅ v1 refreshes, v2 doesn't |

#### 3. **ContentUtilTest** (4 test cases)

Validates content generation and field adapters:

| Scenario                                                     | Result                                |
| ------------------------------------------------------------ | ------------------------------------- |
| Generate correct thread pool content JSON                    | ✅ All required fields included        |
| Field adapter: old/new field names generate identical content | ✅ **Key backward compatibility test** |
| GroupKey generation correctness                              | ✅ Correct format                      |

#### 4. **IncrementalMd5UtilBoundaryTest** (10 test cases)

Boundary conditions and exception handling:

| Scenario                               | Result                              |
| -------------------------------------- | ----------------------------------- |
| null config parameter                  | ✅ Correctly handles NPE             |
| All null fields config                 | ✅ Generates valid MD5               |
| Version 0, negative, very large values | ✅ Correct fallback logic            |
| Config with only extended parameters   | ✅ v2 correctly excludes             |
| Empty string vs null                   | ✅ Correctly differentiates          |
| MD5 calculation consistency            | ✅ Multiple calls return same result |

#### 5. **Server Module Tests** (in `threadpool/server/config`)

**Md5ConfigUtilVersionTest** (8 test cases)

- ✅ Client sends v2 version header - Server correctly identifies
- ✅ Old client without version header - Defaults to v1 (backward compatible)
- ✅ Version header is empty string - Defaults to v1
- ✅ Version header has invalid format - Graceful handling, defaults to v1
- ✅ Version header explicitly v1 - Correctly identified
- ✅ Version header is future version (v3) - Correctly parsed
- ✅ v1 client version detection - compareMd5 will use v1 protocol
- ✅ v2 client version detection - compareMd5 will use v2 protocol

**ConfigCacheServiceVersionTest** (4 test cases)

- ✅ v1 and v2 produce different MD5 (with extended params)
- ✅ v2 client - Extended param change doesn't trigger refresh
- ✅ v1 client - Extended param change triggers refresh (backward compatible)
- ✅ Both versions - Core param change triggers refresh

> **Note:** Server tests focus on version detection logic and MD5 calculation strategy unit tests. Full ConfigCacheService integration tests (requiring configuration cache infrastructure) are part of system integration testing. This PR focuses on core protocol logic unit test coverage,and this part is currently being actively followed up for testing.

#### 6. **CacheDataContentDeliveryTest** (5 test cases)

Validates content delivery layer (fixes content truncation bug):

| Scenario | Result |
| -------- | ------ |
| `setContent` stores full configuration with MD5 separation | ✅ Full content stored, MD5 on core fields |
| Listener receives complete configuration | ✅ All extended params (`executeTimeOut`, `isAlarm`) present |
| MD5 only changes on core parameter change | ✅ Extended param changes don't affect MD5 (v2) |
| **Bug scenario**: Incremental content vs full content | ✅ **Detects pre-fix bug**, validates fix |
| End-to-end: `ServerThreadPoolDynamicRefresh` parsing | ✅ Refresh receives all critical extended params |

> **Critical Fix Validation:** Test 4 demonstrates the bug where using `IncrementalContentUtil.getIncrementalContent()` for content delivery (pre-fix) causes extended parameters to be null. Post-fix uses `ContentUtil.getPoolContent()` to preserve all fields while MD5 calculation still uses incremental content internally.
---

### 2️⃣ Cross-Version Compatibility: **Server 2.0.0 + Client 1.5.0** 

#### Compatibility Matrix

| Server     | Client         | Protocol         | Result           |
| ---------- | -------------- | ---------------- | ---------------- |
| 2.0.0 (v2) | **1.5.0 (v1)** | v1 (full MD5)    | ✅ **Compatible** |
| 2.0.0 (v2) | 2.0.0 (v2)     | v2 (incremental) | ✅ Compatible     |
| 1.5.0 (v1) | 2.0.0 (v2)     | v1 (full MD5)    | ✅ Compatible     |

#### How Server 2.0.0 Supports Client 1.5.0:

**1. Client 1.5.0 behavior:**

- Sends request **without** `X-Hippo4j-Protocol-Version` header
- Uses old field names (`coreSize/maxSize`)

**2. Server 2.0.0 detection:**

```java
// Md5ConfigUtil.java
private static int getClientVersion(HttpServletRequest request) {
    String versionHeader = request.getHeader(PROTOCOL_VERSION_HEADER);
    return StringUtil.isNotEmpty(versionHeader) ? 
           Integer.parseInt(versionHeader) : 1;  // ← Defaults to v1
}
```

**3. Server adaptation:**

```java
// IncrementalMd5Util.java
public static String getVersionedMd5(ThreadPoolParameterInfo config, int clientVersion) {
    if (clientVersion >= 2) {
        return getCoreMd5(config);      // v2 client: incremental
    }
    return getFullMd5(config);          // v1 client: full MD5 ✅
}
```

**4. Field name compatibility:**

```java
// ThreadPoolParameterInfo.java - Field adapter
public Integer corePoolSizeAdapt() {
    return corePoolSize != null ? corePoolSize : coreSize;  // Unified handling
}
```

**Result:** Old v1 clients receive **exactly the same behavior** as before—**zero breaking changes**.

---

### 3️⃣ Implementation Summary

**1. Core Components:**

1. `IncrementalContentUtil` - Defines `PROTOCOL_VERSION=2`, core vs extended parameters
2. `IncrementalMd5Util` - Version-aware MD5 calculation
3. `Md5ConfigUtil` - Client version extraction (defaults to 1)
4. `ConfigCacheService` - Uses versioned MD5 comparison
5. `ClientWorker` - Adds protocol version header
6. `ThreadPoolParameterInfo` - Field adapters for compatibility

**2. Core Parameters (MD5 included):**
`tenantId`, `itemId`, `tpId`, `corePoolSize`, `maximumPoolSize`, `queueType`, `capacity`, `keepAliveTime`, `rejectedType`, `allowCoreThreadTimeOut`

**3. Extended Parameters (MD5 excluded in v2):**
`executeTimeOut`, `isAlarm`, `capacityAlarm`, `livenessAlarm`, `activeAlarm`, `rejectedAlarm`, `alarmInterval`

---

File: starters/threadpool/server/src/main/java/cn/hippo4j/springboot/starter/core/ClientWorker.java [hunk 3]
```diff
@@ -206,6 +207,8 @@ public List<String> checkUpdateTpIds(String probeUpdateString, boolean isInitial
             headers.put(LONG_PULLING_TIMEOUT_NO_HANGUP, "true");
         }
         headers.put(CLIENT_VERSION, version);
+        // Add protocol version header for incremental updates
+        headers.put("X-Hippo4j-Protocol-Version", String.valueOf(IncrementalContentUtil.PROTOCOL_VERSION));
```
Comment (magestacks) [inline starters/threadpool/server/src/main/java/cn/hippo4j/springboot/starter/core/ClientWorker.java]: The client starter version number should not be hardcoded — it should be read from the **pom.xml** file. The current code already includes this line:

```java
headers.put(CLIENT_VERSION, version);
```

After the Server receives the version number, it should determine whether to include the newly added variables in the refresh logic.

For example:
If the Server in version **2.0** introduces a new parameter `xxx`, and the Client version is lower than **2.0**, then this parameter should **not** be included in the refresh check. Similarly, if the Server upgrades to **2.1** and adds new parameters, the same logic should apply.

Currently, the version number is hardcoded as `2` in the code, which does **not** meet the above requirement.

Comment (mingri31164) [reply starters/threadpool/server/src/main/java/cn/hippo4j/springboot/starter/core/ClientWorker.java]: Hi @magestacks ,

Thank you for the detailed feedback! I've addressed the concerns you raised:

## 1. Version Number No Longer Hardcoded ✅

**Before:**

```java
 // hardcoded as 2
headers.put("X-Hippo4j-Protocol-Version", String.valueOf(IncrementalContentUtil.PROTOCOL_VERSION));
```

**After:**

```java
// ClientWorker.java:97-99
this.version = VersionUtil.resolveClientVersion(version, ClientWorker.class);
int resolvedProtocol = VersionUtil.resolveProtocolVersion(this.version, IncrementalContentUtil.PROTOCOL_VERSION);
this.protocolVersion = VersionUtil.UNKNOWN_VERSION.equals(this.version) 
    ? IncrementalContentUtil.PROTOCOL_VERSION 
    : resolvedProtocol;

// ClientWorker.java:214-216
headers.put(CLIENT_VERSION, version); // from pom.xml via BuildProperties
headers.put("X-Hippo4j-Protocol-Version", String.valueOf(protocolVersion)); // dynamically resolved
```

The client version is now read from `pom.xml` through `BuildProperties`, and the protocol version is dynamically mapped using `VersionUtil`:

- Version 1.x → Protocol v1 (full MD5)
- Version 2.x → Protocol v2 (incremental MD5)
- Future versions can be registered in `VersionUtil.PROTOCOL_VERSION_MAPPINGS`

---

## 2. Server Determines Field Inclusion Based on Client Version ✅

**Implementation:**

I've introduced a field-level version control mechanism via `FIELD_MIN_PROTOCOL_VERSION`:

```java
// IncrementalContentUtil.java
private static final Map<String, Integer> FIELD_MIN_PROTOCOL_VERSION;

static {
    Map<String, Integer> fieldVersion = new HashMap<>();
    IDENTIFIER_FIELDS.forEach(field -> fieldVersion.put(field, 1));
    CORE_PARAMETER_LIST.forEach(field -> fieldVersion.put(field, 1));
    EXTENDED_PARAMETER_LIST.forEach(field -> fieldVersion.put(field, PROTOCOL_VERSION + 1));
    FIELD_MIN_PROTOCOL_VERSION = Collections.unmodifiableMap(fieldVersion);
}

private static boolean shouldIncludeField(String field, int protocolVersion) {
    int minProtocol = FIELD_MIN_PROTOCOL_VERSION.getOrDefault(field, Integer.MAX_VALUE);
    return protocolVersion >= minProtocol;
}
```

**How it works:**

When the Server receives a client's version number, it:

1. Parses `X-Hippo4j-Protocol-Version` header or infers from `Client-Version`
2. Generates versioned content by filtering fields based on `FIELD_MIN_PROTOCOL_VERSION`
3. Only includes fields whose minimum protocol version ≤ client's protocol version

---

## 3. Example Scenarios (As You Described) ✅

### Scenario 1: Server 2.0 introduces parameter `xxx`

**Configuration:**

```java
FIELD_MIN_PROTOCOL_VERSION.put("xxx", 2); // requires protocol v2
```

**Client Behavior:**

- **Client v1.9 (protocol v1)**: Uses full MD5 comparison (backward compatible)
- **Client v2.0+ (protocol v2)**: 
  - If `xxx` is configured for protocol v2 → includes `xxx` in MD5
  - If `xxx` is configured for protocol v3 → excludes `xxx` from MD5 ✅

### Scenario 2: Server 2.1 adds new parameter `yyy`

**Configuration:**

```java
FIELD_MIN_PROTOCOL_VERSION.put("yyy", 3); // requires protocol v3
```

**Client Behavior:**

- **Client v2.0 (protocol v2)**: Excludes `yyy` from MD5 → no refresh when `yyy` changes ✅
- **Client v2.1 (protocol v2)**: Same as v2.0 (still protocol v2)
- **Client v3.0 (protocol v3)**: Includes `yyy` in MD5 → refreshes when `yyy` changes ✅

---

## 4. Testing & Verification ✅

I've added comprehensive test coverage:

**New Test Class: `FieldVersionControlTest`** (8 test cases)

- `testMentorScenario_ServerV20AddsFieldXxx`: Validates your 2.0 example ✅
- `testMentorScenario_ServerV21AddsFieldYyy`: Validates your 2.1 example ✅
- `testFieldVisibilityMatrix`: Shows field visibility across protocol versions ✅

Comment (mingri31164) [reply starters/threadpool/server/src/main/java/cn/hippo4j/springboot/starter/core/ClientWorker.java]: Hi @magestacks ,

Thank you for the detailed feedback! I've refactored the implementation to address all your concerns in commit d453c37. Here's a comprehensive summary of the changes:

---

## 🎯 Key Changes

### 1. Eliminated Hardcoded Version Numbers ✅

**Before:**
```java
// Hardcoded as 2
headers.put("X-Hippo4j-Protocol-Version", String.valueOf(IncrementalContentUtil.PROTOCOL_VERSION)); 
```

**After:**
```java
// ClientWorker.java - Dynamic version resolution
this.version = VersionUtil.resolveClientVersion(version, ClientWorker.class);
// Reads from BuildProperties (pom.xml) → e.g., "2.0.1"

int resolvedProtocol = VersionUtil.resolveProtocolVersion(this.version, IncrementalContentUtil.PROTOCOL_VERSION);
this.protocolVersion = VersionUtil.UNKNOWN_VERSION.equals(this.version) 
    ? IncrementalContentUtil.PROTOCOL_VERSION 
    : resolvedProtocol;
// "1.5.0" → protocol 1, "2.0.1" → protocol 2

headers.put(CLIENT_VERSION, version); // from pom.xml
headers.put("X-Hippo4j-Protocol-Version", String.valueOf(protocolVersion)); // dynamically calculated
```

---

### 2. Implemented Dynamic Field-to-Version Binding ✅

**Core Mechanism:**

I've introduced a **runtime metadata system** that allows Server to dynamically bind fields to semantic versions without code changes:

```java
// IncrementalFieldMetadataProvider.java - New interface
public interface IncrementalFieldMetadataProvider {
    default Map<String, String> getFieldVersionMetadata() {
        return Collections.emptyMap();
    }
}

// Server-side configuration
ConfigAllInfo config = new ConfigAllInfo();
Map<String, String> metadata = new HashMap<>();
metadata.put("executeTimeOut", "2.0.0");  // Field A introduced in Server 2.0
metadata.put("capacityAlarm", "2.1.0");   // Field B introduced in Server 2.1
config.setFieldVersionMetadata(metadata);
```

**How it works:**
1. Server declares field versions via `setFieldVersionMetadata()`
2. Metadata is automatically serialized into JSON
3. Client receives and parses metadata automatically
4. Both sides use identical filtering logic based on semantic version comparison

---

### 3. Semantic Version-Based Field Filtering ✅

**Implementation:**

```java
// VersionUtil.java - New method
public static boolean isVersionGreaterOrEqual(String version1, String version2) {
    SemanticVersion v1 = SemanticVersion.parse(version1);
    SemanticVersion v2 = SemanticVersion.parse(version2);
    return v1.compareTo(v2) >= 0;
    // Supports: "2.0.5" >= "2.0.0", "2.0.5" < "2.1.0"
}

// IncrementalContentUtil.java - Field filtering logic
private static boolean shouldIncludeField(String field, String clientVersion, Map<String, String> fieldRules) {
    String minVersion = fieldRules.get(field);
    return VersionUtil.isVersionGreaterOrEqual(clientVersion, minVersion);
}
```

---

### 4. Your Examples Now Work Exactly as Specified ✅

#### Example 1: Server 2.0 introduces field `executeTimeOut`

**Configuration:**
```java
metadata.put("executeTimeOut", "2.0.0");
```

**Client Behavior:**

| Client Version | Version Comparison | Includes executeTimeOut? | Refreshes on Change? |
|---------------|-------------------|-------------------------|---------------------|
| 1.5.0 | `"1.5.0" >= "2.0.0"` → false | ❌ No | ❌ No (prevents invalid refresh) ✅ |
| 1.9.9 | `"1.9.9" >= "2.0.0"` → false | ❌ No | ❌ No ✅ |
| 2.0.0 | `"2.0.0" >= "2.0.0"` → true | ✅ Yes | ✅ Yes (as expected) |
| 2.0.5 | `"2.0.5" >= "2.0.0"` → true | ✅ Yes | ✅ Yes |

---

#### Example 2: Server 2.1 introduces field `capacityAlarm`

**Configuration:**
```java
metadata.put("capacityAlarm", "2.1.0");
```

**Client Behavior:**

| Client Version | Includes executeTimeOut (2.0.0)? | Includes capacityAlarm (2.1.0)? | 
|---------------|----------------------------------|----------------------------------|
| 1.5.0 | ❌ No | ❌ No |
| 2.0.0 | ✅ Yes | ❌ No (prevents invalid refresh) ✅ |
| 2.0.9 | ✅ Yes | ❌ No ✅ |
| 2.1.0 | ✅ Yes | ✅ Yes |

**Review by magestacks — APPROVED** [review]
